File: 精选TOP 面试题/5. 最长回文子串.py
# class Solution:
#     def longestPalindrome(self, s):
#         if len(s) <= 1:
#             return s
#         result = {}
#         for i in range(len(s)):
#             for j in range(i+1, len(s)+1):
#                 substring = s[i:j]
#                 rs = substring[::-1]
#                 if substring == rs:
#                     result[substring] = len(rs)
#         r = sorted(result.items(),key=lambda x:x[1], reverse=True)
#         return r[0][0]

class Solution:
    def longestPalindrome(self, s):
        if len(s) <= 1:
            return s
        size = len(s)
        dp = [[False for _ in range(size)] for _ in range(size)]
        longest_1 = 1
        res = s[0]
        for r in range(1, size):
            for l in range(r):
                if s[l]==s[r] and (r-l<=2 or dp[l+1][r-1]):
                    dp[l][r] = True
                    cur_len=r-l+1
                    if cur_len>longest_1:
                        longest_1=cur_len
                        res=s[l:r+1]
        return res

string = "ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggg"
# string = "babad"
s = Solution()
r=s.longestPalindrome(string)
print(r)

File: 精选TOP 面试题/5.最长回文子串.cpp
#include<iostream>
#include<string> 
#include<vector>
using namespace std; 
class Solution {
public:
    string longestPalindrome(string s) {
        if (s.length()<=1){
                return s;
        }
        int sizes = s.length();
        int cur_len = 1;
        int longest_l = 1;
        string res = s.substr(0,1);
        vector<vector<bool> > dp(sizes,vector<bool>(sizes,false));
        for(int r=1;r<sizes;r++){
            for(int l=0;l<r;l++){
                if ((s[l]==s[r])&((r-l<=2)|(dp[l+1][r-1]))){
                    dp[l][r] = true;
                    cur_len = r-l+1;
                    if (cur_len>longest_l){
                        longest_l = cur_len;
                        res = s.substr(l, r-l+1);
                    }
                }
            }
        }
        return res;
    }
};

int main(void){
	string strs = "babad";
	Solution s = Solution();
	string r =s.longestPalindrome(strs);
	cout<<r<<endl;
	return 0;
}
